openvpncloud/resource_service.go

```diff
@@ -0,0 +1,307 @@
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"description": {
+				Type: schema.TypeString,
+				ValidateDiagFunc: func(i interface{}, path cty.Path) diag.Diagnostics {
+					val := i.(string)
+					if len(val) > 255 {
+						return diag.Errorf("description must be less than 255 characters")
+					}
+					if len(val) == 0 {
+						return diag.Errorf("description must not be empty")
+					}
+					return nil
+				},
+				Optional: true,
+			},
+			"type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateDiagFunc: func(i interface{}, path cty.Path) diag.Diagnostics {
+					val := i.(string)
+					if val != "IP_SOURCE" && val != "SERVICE_DESTINATION" {
+						return diag.Errorf("type must be one of IP_SOURCE or SERVICE_DESTINATION")
+					}
+					return nil
+				},
+			},
+			// list of strings
+			"routes": {
+				Type:     schema.TypeList,
+				Required: true,
+				MinItems: 1,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"config": {
+				Type:     schema.TypeList,
+				MaxItems: 1,
+				Optional: true,
+				Elem:     resourceServiceConfig(),
+				ForceNew: true, //remove
+			},
+			"network_item_type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: func(i interface{}, s string) ([]string, []error) {
+					if i.(string) != "NETWORK" && i.(string) != "HOST" {
+						return nil, []error{fmt.Errorf("network_item_type must be one of NETWORK or HOST")}
+					}
+					return nil, nil
+				},
+			},
+			"network_item_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourceServiceUpdate(ctx context.Context, data *schema.ResourceData, i interface{}) diag.Diagnostics {
+	c := i.(*client.Client)
+	networkItemId := data.Get("network_item_id").(string)
+	networkItemType := data.Get("network_item_type").(string)
+
+	s, err := c.UpdateService(data.Id(), networkItemType, networkItemId, resourceDataToService(data))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	setResourceData(data, s)
+	return nil
+}
+
+func resourceServiceConfig() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"custom_service_types": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"icmp_type": {
+							Type:     schema.TypeList,
+							Required: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"lower_value": {
+										Type:     schema.TypeInt,
+										Required: true,
+									},
+									"upper_value": {
+										Type:     schema.TypeInt,
+										Required: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			"service_types": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+					ValidateDiagFunc: func(i interface{}, path cty.Path) diag.Diagnostics {
```

Comment:
The `validation` library has a built-in function for checking that the value is in a string slice https://github.com/hashicorp/terraform-plugin-sdk/blob/main/helper/validation/strings.go#LL132C67-L132C67 (unless you want to have a very specific error message).

Reply:
You can use that same function in the other fields where you're explicitly checking for values

Reply:
thanks! replaced with `validation.StringInSlice`